src/compare-tree/index.js

```diff
@@ -0,0 +1,193 @@
+   * Create a subscription to a particular set of key paths.
+   * @public
+   * @param {String} keyPaths A list of key paths to trigger the callback
+   * @param {Function} callback Function to execute when a key path changes
+   * @param {*} [scope] Optional scope to invoke the function with
+   */
+  on (keyPaths, callback, scope) {
+    let dependencies = getKeyPaths(keyPaths)
+    let id = Query.getId(keyPaths)
+
+    let query = this.addQuery(id, dependencies)
+
+    dependencies.map(this.track, this)
+                .forEach(node => node.connect(query))
+
+    query.on('change', callback, scope)
+
+    return query
+  },
+
+  /**
+   * Remove a subscription created by .on()
+   * @public
+   * @param {String} keyPaths A list of key paths for the subscription
+   * @param {Function} callback Associated callback function
+   * @param {*} [scope] Associated scope
+   */
+  off (keyPaths, callback, scope) {
+    let id = Query.getId(keyPaths)
+
+    let query = this.nodes[id]
+
+    if (query) {
+      query.off('change', callback, scope)
+
+      if (query.isEmpty()) {
+        this.prune(query)
+      }
+    }
+  },
+
+  /**
+   * Compare a new snapshot to the last snapshot, triggering event
+   * subscriptions along the way.
+   * @public
+   * @param {*} state - New snapshot of state
+   */
+  update (snapshot) {
+    let last = this.snapshot
+    this.snapshot = snapshot
+
+    let root = this.nodes[ROOT_KEY]
+
+    if (root) {
+      this.scan(root, last, snapshot)
+    }
+  },
+
+  /**
+   * Add a node to the tree if it has not otherwise been added.
+   * @private
+   * @param {String} id Identifier for the node.
+   * @param {String} key Each node represents a key in a nested object.
+   * @param {Node} parent Parent to connect this node to.
+   */
+  addNode (id, key, parent) {
+    if (!this.nodes[id]) {
+      this.nodes[id] = new Node(id, key, parent)
+    }
+
+    return this.nodes[id]
+  },
+
+  /**
+   * Add a query to the tree if it has not otherwise been
+   * added. Queries are leaf nodes responsible for managing
+   * subscriptions.
+   * @private
+   * @param {String} id Identifier for the node.
+   * @param {String} keyPaths Each query tracks a list of key paths
+   */
+  addQuery (id, keyPaths) {
+    if (!this.nodes[id]) {
+      this.nodes[id] = new Query(id, keyPaths)
+    }
+
+    return this.nodes[id]
+  },
+
+  /**
+   * Remove a query, then traverse that queries key paths to remove
+   * unused parents.
+   * @private
+   * @param {Query} query Query to remove
+   */
+  prune (query) {
+    let { keyPaths } = query
+
+    for (var i = 0, len = keyPaths.length; i < len; i++) {
+      let path = keyPaths[i]
+      let id = path.length ? getKeyString(keyPaths[i]) : ROOT_KEY
+      let node = this.nodes[id]
+
+      node.disconnect(query)
+
+      do {
+        if (node.isAlone()) {
+          node.orphan()
+          delete this.nodes[node.id]
+        }
+
+        node = node.parent
+      } while (node)
+    }
```

Comment:
This is starting to feel better, but I don't like that there's an exception to the rules for how IDs are generated for the root node. Basically, the root node has an id of `'~'` to work around the fact that it would otherwise by an empty string (`''`), which they key path utilities convert into an empty array.

I think this exception could be eliminated. Just need to spend some more time on it.